tests/system_tests/test_core/test_metric_full.py

```diff
@@ -308,96 +223,78 @@ def test_metric_goal(relay_server, wandb_init):
+    with wandb_backend_spy.freeze() as snapshot:
+        summary = snapshot.summary(run_id=run.id)
+        assert math.isnan(summary["val"]["min"])
 
 
 @pytest.mark.wandb_core_only(
     reason="deviates from legacy behavior as nan value should be respected"
 )
-def test_metric_nan_min_more(relay_server, wandb_init):
-    with relay_server() as relay:
-        run = wandb_init()
-        run_id = run.id
+def test_metric_nan_min_more(wandb_backend_spy):
+    with wandb.init() as run:
         run.define_metric("val", summary="min")
         run.log(dict(mystep=1, val=float("nan")))
         run.log(dict(mystep=1, val=4))
-        run.finish()
-
-    summary = relay.context.get_run_summary(run_id)
 
-    assert math.isnan(summary["val"]["min"])
+    with wandb_backend_spy.freeze() as snapshot:
+        summary = snapshot.summary(run_id=run.id)
+        assert math.isnan(summary["val"]["min"])
 
 
-def test_metric_nested_default(relay_server, wandb_init):
-    with relay_server() as relay:
-        run = wandb_init()
-        run_id = run.id
+def test_metric_nested_default(wandb_backend_spy):
+    with wandb.init() as run:
         run.log(dict(this=dict(that=3)))
         run.log(dict(this=dict(that=2)))
         run.log(dict(this=dict(that=4)))
-        run.finish()
 
-    summary = relay.context.get_run_summary(run_id)
-
-    assert summary["this"] == {"that": 4}
+    with wandb_backend_spy.freeze() as snapshot:
+        summary = snapshot.summary(run_id=run.id)
+        assert summary["this"] == {"that": 4}
 
 
-def test_metric_nested_copy(relay_server, wandb_init):
-    with relay_server() as relay:
-        run = wandb_init()
-        run_id = run.id
+def test_metric_nested_copy(wandb_backend_spy):
+    with wandb.init() as run:
         run.define_metric("this.that", summary="copy")
         run.log(dict(this=dict(that=3)))
         run.log(dict(this=dict(that=2)))
         run.log(dict(this=dict(that=4)))
-        run.finish()
-
-    summary = relay.context.get_run_summary(run_id)
 
-    assert summary["this"] == {"that": 4}
+    with wandb_backend_spy.freeze() as snapshot:
+        summary = snapshot.summary(run_id=run.id)
+        assert summary["this"] == {"that": 4}
 
 
-def test_metric_nested_min(relay_server, wandb_init):
-    with relay_server() as relay:
-        run = wandb_init()
-        run_id = run.id
+def test_metric_nested_min(wandb_backend_spy):
+    with wandb.init() as run:
         run.define_metric("this.that", summary="min")
         run.log(dict(this=dict(that=3)))
         run.log(dict(this=dict(that=2)))
         run.log(dict(this=dict(that=4)))
-        run.finish()
-
-    summary = relay.context.get_run_summary(run_id)
 
-    assert summary["this"] == {"that": {"min": 2}}
+    with wandb_backend_spy.freeze() as snapshot:
+        summary = snapshot.summary(run_id=run.id)
+        assert summary["this"] == {"that": {"min": 2}}
 
 
 def test_metric_nested_mult(relay_server, wandb_init):
```

Comment:
is this still using relay_server because you need to implement the logic that capture metrics?

Reply:
yes! I wanted to limit the initial PR's size